### PR TITLE
Admin APIs and tools for node params (without refreshing)

### PR DESCRIPTION
**Pull request by kriii**
Changes
- Generalize exist APIs
- Fix tools due to API changes
- Fix some typos

TODO
- Make the collection of callbacks for some parameters that need a refresh.

Related Issue
- #956

### CODE REVIEW

File: tools/api-access/removeFromDevClientApiIpWhitelist.js [hunk 1]
```diff
@@ -61,9 +62,9 @@ async function processArguments() {
 function usage() {
   console.log('\nUsage:\n  node removeFromDevClientApiIpWhitelist.js <NODE_ENDPOINT> <CHAIN_ID> <ACCOUNT_TYPE> [<KEYSTORE_FILE_PATH>] <IP_ADDRESS>\n');
   console.log('\nExamples:');
-  console.log('node tools/api-access/getDevClientApiIpWhitelist.js http://localhost:8081 0 private_key 127.0.0.1');
-  console.log('node tools/api-access/getDevClientApiIpWhitelist.js http://localhost:8081 0 mnemonic 127.0.0.1');
-  console.log('node tools/api-access/getDevClientApiIpWhitelist.js http://localhost:8081 0 keystore /path/to/keystore/file 127.0.0.1');
+  console.log('node tools/api-access/removeFromDevClientApiIpWhitelist.js http://localhost:8081 0 private_key 127.0.0.1');
```
Comment (liayoo) [inline tools/api-access/removeFromDevClientApiIpWhitelist.js:65]: thanks for fixing these~ 👍

File: common/result-code.js [hunk 2]
```diff
@@ -173,15 +173,15 @@ const JsonRpcApiResultCode = {
...
-  IP_ALREADY_IN_WHITELIST: 30502,
-  // ain_removeFromDevClientApiIpWhitelist
-  IP_NOT_IN_WHITELIST: 30601,
+  // Admin APIs
+  PARAM_INVALID: 30501,
+  VALUE_INCOMPATIBLE: 30502,
+  ALREADY_IN_WHITELIST: 30503,
+  NOT_IN_WHITELIST: 30504,
   // ain_validateAppName
-  INVALID_APP_NAME_FOR_STATE_LABEL: 30701,
```
Comment (liayoo) [inline common/result-code.js:182]: Could you make sure to run the unit & integration tests? I think they would need some updates from this change.

File: json_rpc/constants.js [hunk 3]
```diff
@@ -13,11 +13,11 @@ const JSON_RPC_METHODS = {
   AIN_GET_BLOCK_TRANSACTION_COUNT_BY_HASH: 'ain_getBlockTransactionCountByHash',
   AIN_GET_BLOCK_TRANSACTION_COUNT_BY_NUMBER: 'ain_getBlockTransactionCountByNumber',
   AIN_GET_BOOTSTRAP_PUB_KEY: 'ain_getBootstrapPubKey',
-  AIN_GET_DEV_CLIENT_API_IP_WHITELIST: 'ain_getDevClientApiIpWhitelist',
   AIN_GET_EVENT_HANDLER_CHANNEL_INFO: 'ain_getEventHandlerChannelInfo',
   AIN_GET_EVENT_HANDLER_FILTER_INFO: 'ain_getEventHandlerFilterInfo',
   AIN_GET_LAST_BLOCK: 'ain_getLastBlock',
   AIN_GET_LAST_BLOCK_NUMBER: 'ain_getLastBlockNumber',
+  AIN_GET_NODE_PARAM: 'ain_getNodeParam',
```
Comment (liayoo) [inline json_rpc/constants.js:20]: Could you add tool scripts for the new apis?

File: json_rpc/admin.js [hunk 4]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+      }
+
+      const param = args.message.param;
       const latency = Date.now() - beginTime;
-      trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_GET, latency);
-      if (_.get(args.message, 'method') === JSON_RPC_METHODS.AIN_GET_DEV_CLIENT_API_IP_WHITELIST &&
-          verified) {
-        done(null,
-            JsonRpcUtil.addProtocolVersion({ result: NodeConfigs.DEV_CLIENT_API_IP_WHITELIST }));
+      trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
```
Comment (platfowner) [inline json_rpc/admin.js:39]: s/ACCESS_CONTROL_SET/ACCESS_CONTROL_GET/?

File: json_rpc/admin.js [hunk 5]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+module.exports = function getAdminApis(node) {
   return {
-    [JSON_RPC_METHODS.AIN_GET_DEV_CLIENT_API_IP_WHITELIST]: function(args, done) {
+    [JSON_RPC_METHODS.AIN_GET_NODE_PARAM]: function(args, done) {
       const beginTime = Date.now();
       const verified = node.verifyNodeAccountSignature(args.message, args.signature);
+      if (_.get(args.message, 'method') !== JSON_RPC_METHODS.AIN_GET_NODE_PARAM || !verified) {
+        const latency = Date.now() - beginTime;
+        trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
+        done({ code: 403, message: 'Forbidden' });
```
Comment (platfowner) [inline json_rpc/admin.js:33]: it looks a little bit out of the scope but could you define a result code in JsonRpcApiResultCode for this case?

File: json_rpc/admin.js [hunk 6]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
-    [JSON_RPC_METHODS.AIN_ADD_TO_DEV_CLIENT_API_IP_WHITELIST]: function(args, done) {
+    [JSON_RPC_METHODS.AIN_SET_NODE_PARAM]: function(args, done) {
       const beginTime = Date.now();
       const verified = node.verifyNodeAccountSignature(args.message, args.signature);
-      if (_.get(args.message, 'method') !== JSON_RPC_METHODS.AIN_ADD_TO_DEV_CLIENT_API_IP_WHITELIST ||
-          !verified) {
+      if (_.get(args.message, 'method') !== JSON_RPC_METHODS.AIN_SET_NODE_PARAM || !verified) {
         const latency = Date.now() - beginTime;
         trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
         done({ code: 403, message: 'Forbidden' });
```
Comment (platfowner) [inline json_rpc/admin.js:58]: ditto?

File: json_rpc/admin.js [hunk 7]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+      }));
+    },
+
+    [JSON_RPC_METHODS.AIN_ADD_TO_WHITELIST_NODE_PARAM]: function(args, done) {
+      const beginTime = Date.now();
+      const verified = node.verifyNodeAccountSignature(args.message, args.signature);
+      if (_.get(args.message, 'method') !== JSON_RPC_METHODS.AIN_ADD_TO_WHITELIST_NODE_PARAM || !verified) {
+        const latency = Date.now() - beginTime;
+        trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
+        done({ code: 403, message: 'Forbidden' });
```
Comment (platfowner) [inline json_rpc/admin.js:104]: ditto?

File: json_rpc/admin.js [hunk 8]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
-      if (_.get(args.message, 'method') === JSON_RPC_METHODS.AIN_GET_DEV_CLIENT_API_IP_WHITELIST &&
-          verified) {
-        done(null,
-            JsonRpcUtil.addProtocolVersion({ result: NodeConfigs.DEV_CLIENT_API_IP_WHITELIST }));
+      trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
+      if (NodeConfigs[param] === undefined) {
+        done(null, JsonRpcUtil.addProtocolVersion({
+          result: {
+            code: JsonRpcApiResultCode.PARAM_INVALID,
+            message: `Param [${param}] is not exist.`
```
Comment (platfowner) [inline json_rpc/admin.js:44]: s/is/does/?

File: json_rpc/admin.js [hunk 9]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+      const param = args.message.param;
+      if (NodeConfigs[param] === undefined) {
         const latency = Date.now() - beginTime;
         trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
         done(null, JsonRpcUtil.addProtocolVersion({
           result: {
-            code: JsonRpcApiResultCode.SUCCESS,
-            message: `Added IP (${args.message.ip}) to whitelist: ${JSON.stringify(NodeConfigs.DEV_CLIENT_API_IP_WHITELIST)}`
+            code: JsonRpcApiResultCode.PARAM_INVALID,
+            message: `Param [${param}] is not exists.`
```
Comment (platfowner) [inline json_rpc/admin.js:69]: ditto?

File: json_rpc/admin.js [hunk 10]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+      const param = args.message.param;
+      if (NodeConfigs[param] === undefined) {
         const latency = Date.now() - beginTime;
         trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
         done(null, JsonRpcUtil.addProtocolVersion({
           result: {
-            code: JsonRpcApiResultCode.IP_ALREADY_IN_WHITELIST,
-            message: `IP (${args.message.ip}) already in whitelist: ${JSON.stringify(NodeConfigs.DEV_CLIENT_API_IP_WHITELIST)}`
+            code: JsonRpcApiResultCode.PARAM_INVALID,
+            message: `Param [${param}] is not exists.`
```
Comment (platfowner) [inline json_rpc/admin.js:115]: s/is not exists/does not exist/?

File: json_rpc/admin.js [hunk 11]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+      const param = args.message.param;
+      if (NodeConfigs[param] === undefined) {
         const latency = Date.now() - beginTime;
         trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
         done(null, JsonRpcUtil.addProtocolVersion({
           result: {
-            code: JsonRpcApiResultCode.SUCCESS,
-            message: `Removed IP (${args.message.ip}) from whitelist: ${JSON.stringify(NodeConfigs.DEV_CLIENT_API_IP_WHITELIST)}`
+            code: JsonRpcApiResultCode.PARAM_INVALID,
+            message: `Param [${param}] is not exists.`
```
Comment (platfowner) [inline json_rpc/admin.js:181]: ditto?

File: json_rpc/admin.js [hunk 12]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+        return;
+      }
+      if (!CommonUtil.isArray(NodeConfigs[param]) &&
+        !CommonUtil.isWildcard(NodeConfigs[param])) {
+        const latency = Date.now() - beginTime;
+        trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
+        done(null, JsonRpcUtil.addProtocolVersion({
+          result: {
+            code: JsonRpcApiResultCode.PARAM_INVALID,
+            message: `Param [${param}] is not whitelist`
```
Comment (platfowner) [inline json_rpc/admin.js:127]: s/not whitelist/not in whitelist/?

File: common/result-code.js [hunk 13]
```diff
@@ -173,15 +173,15 @@ const JsonRpcApiResultCode = {
   BATCH_TX_MISSING_PROPERTIES: 30404,
   BATCH_TX_INVALID_FORMAT: 30405,
   BATCH_TX_INVALID_SIGNATURE: 30406,
-  // ain_addToDevClientApiIpWhitelist
-  INVALID_IP: 30501,
-  IP_ALREADY_IN_WHITELIST: 30502,
-  // ain_removeFromDevClientApiIpWhitelist
-  IP_NOT_IN_WHITELIST: 30601,
+  // Admin APIs
+  PARAM_INVALID: 30501,
```
Comment (platfowner) [inline common/result-code.js:177]: suggestion) what about adding prefix 'ADMIN_' for these admin API code?

File: json_rpc/admin.js [hunk 14]
```diff
@@ -10,124 +9,215 @@ const CommonUtil = require('../common/common-util');
...
+      if (!CommonUtil.isArray(NodeConfigs[param]) &&
+        !CommonUtil.isWildcard(NodeConfigs[param])) {
         const latency = Date.now() - beginTime;
         trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
         done(null, JsonRpcUtil.addProtocolVersion({
           result: {
-            code: JsonRpcApiResultCode.IP_NOT_IN_WHITELIST,
-            message: `IP (${args.message.ip}) not in whitelist: ${JSON.stringify(NodeConfigs.DEV_CLIENT_API_IP_WHITELIST)}`
+            code: JsonRpcApiResultCode.PARAM_INVALID,
+            message: `Param [${param}] is not whitelist`
```
Comment (platfowner) [inline json_rpc/admin.js:193]: ditto?

File: tools/api-access/addToDevClientApiIpWhitelist.js [hunk 15]
```diff
@@ -61,9 +62,9 @@ async function processArguments() {
 function usage() {
   console.log('\nUsage:\n  node addToDevClientApiIpWhitelist.js <NODE_ENDPOINT> <CHAIN_ID> <ACCOUNT_TYPE> [<KEYSTORE_FILE_PATH>] <IP_ADDRESS>\n');
   console.log('\nExamples:');
-  console.log('node tools/api-access/getDevClientApiIpWhitelist.js http://localhost:8081 0 private_key 127.0.0.1');
-  console.log('node tools/api-access/getDevClientApiIpWhitelist.js http://localhost:8081 0 mnemonic 127.0.0.1');
-  console.log('node tools/api-access/getDevClientApiIpWhitelist.js http://localhost:8081 0 keystore /path/to/keystore/file 127.0.0.1');
+  console.log('node tools/api-access/addToDevClientApiIpWhitelist.js http://localhost:8081 0 private_key 127.0.0.1');
```
Comment (platfowner) [inline tools/api-access/addToDevClientApiIpWhitelist.js:65]: thanks for fixing this!

File: tools/api-access/removeFromDevClientApiIpWhitelist.js [hunk 16]
```diff
@@ -61,9 +62,9 @@ async function processArguments() {
...
+  console.log('node tools/api-access/removeFromDevClientApiIpWhitelist.js http://localhost:8081 0 mnemonic 127.0.0.1');
+  console.log('node tools/api-access/removeFromDevClientApiIpWhitelist.js http://localhost:8081 0 keystore /path/to/keystore/file 127.0.0.1');
```
Comment (platfowner) [inline tools/api-access/removeFromDevClientApiIpWhitelist.js:67]: what about adding an example with wildcard ('*')?

File: tools/api-access/addToDevClientApiIpWhitelist.js [hunk 17]
```diff
@@ -61,9 +62,9 @@ async function processArguments() {
...
+  console.log('node tools/api-access/addToDevClientApiIpWhitelist.js http://localhost:8081 0 mnemonic 127.0.0.1');
+  console.log('node tools/api-access/addToDevClientApiIpWhitelist.js http://localhost:8081 0 keystore /path/to/keystore/file 127.0.0.1');
```
Comment (platfowner) [inline tools/api-access/addToDevClientApiIpWhitelist.js:67]: what about adding an example with wildcard('*')?

File: json_rpc/admin.js (hunk 4 above)
Comment (kriii) [reply json_rpc/admin.js]: Didn't notice it! Fixed now, thanks!

File: json_rpc/admin.js (hunk 5 above)
Comment (kriii) [reply json_rpc/admin.js]: Added!

File: json_rpc/admin.js (hunk 6 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: json_rpc/admin.js (hunk 7 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: json_rpc/admin.js (hunk 8 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: json_rpc/admin.js (hunk 9 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: json_rpc/admin.js (hunk 10 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: json_rpc/admin.js (hunk 11 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: json_rpc/admin.js (hunk 12 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: common/result-code.js (hunk 13 above)
Comment (kriii) [reply common/result-code.js]: Added!

File: json_rpc/admin.js (hunk 14 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: tools/api-access/removeFromDevClientApiIpWhitelist.js (hunk 16 above)
Comment (kriii) [reply tools/api-access/removeFromDevClientApiIpWhitelist.js:67]: Added!

File: tools/api-access/addToDevClientApiIpWhitelist.js (hunk 17 above)
Comment (kriii) [reply tools/api-access/addToDevClientApiIpWhitelist.js:67]: Added!

File: json_rpc/constants.js (hunk 3 above)
Comment (kriii) [reply json_rpc/constants.js:20]: New tools added!

**Comment by kriii** [comment]
I addressed all reviews and made some changes in API, please take a look!

File: json_rpc/admin.js [hunk 18]
```diff
@@ -10,124 +9,253 @@ const CommonUtil = require('../common/common-util');
...
+      const param = args.message.param;
+      if (NodeConfigs[param] === undefined) {
         const latency = Date.now() - beginTime;
         trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
         done(null, JsonRpcUtil.addProtocolVersion({
           result: {
-            code: JsonRpcApiResultCode.SUCCESS,
-            message: `Added IP (${args.message.ip}) to whitelist: ${JSON.stringify(NodeConfigs.DEV_CLIENT_API_IP_WHITELIST)}`
+            code: JsonRpcApiResultCode.ADMIN_PARAM_INVALID,
+            message: `Param [${param}] does not exists.`
```
Comment (platfowner) [inline json_rpc/admin.js:78]: minor) s/exists/exist/?

File: json_rpc/admin.js [hunk 19]
```diff
@@ -10,124 +9,253 @@ const CommonUtil = require('../common/common-util');
...
+      const param = args.message.param;
+      if (NodeConfigs[param] === undefined) {
         const latency = Date.now() - beginTime;
         trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
         done(null, JsonRpcUtil.addProtocolVersion({
           result: {
-            code: JsonRpcApiResultCode.IP_ALREADY_IN_WHITELIST,
-            message: `IP (${args.message.ip}) already in whitelist: ${JSON.stringify(NodeConfigs.DEV_CLIENT_API_IP_WHITELIST)}`
+            code: JsonRpcApiResultCode.ADMIN_PARAM_INVALID,
+            message: `Param [${param}] does not exists.`
```
Comment (platfowner) [inline json_rpc/admin.js:129]: ditto

File: json_rpc/admin.js [hunk 20]
```diff
@@ -10,124 +9,253 @@ const CommonUtil = require('../common/common-util');
...
-        NodeConfigs.DEV_CLIENT_API_IP_WHITELIST = [];
+
+      const param = args.message.param;
+      if (NodeConfigs[param] === undefined) {
+        const latency = Date.now() - beginTime;
+        trafficStatsManager.addEvent(TrafficEventTypes.ACCESS_CONTROL_SET, latency);
+        done(null, JsonRpcUtil.addProtocolVersion({
+          result: {
+            code: JsonRpcApiResultCode.ADMIN_PARAM_INVALID,
+            message: `Param [${param}] does not exists.`
```
Comment (platfowner) [inline json_rpc/admin.js:207]: ditto

File: json_rpc/admin.js (hunk 18 above)
Comment (kriii) [reply json_rpc/admin.js]: Sorry, fixed!

File: json_rpc/admin.js (hunk 19 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

File: json_rpc/admin.js (hunk 20 above)
Comment (kriii) [reply json_rpc/admin.js]: Fixed!

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

**Review by liayoo — APPROVED** [review]
LGTM, thanks!!

**Comment by kriii** [comment]
Now merged! Thanks for the reviews!

**Pull request merged**